[IM] Index templates form wizard

alisonelizabeth opened the pull request:
This PR adds the ability to create, edit and clone index templates. Includes API and component integration tests, as well as a simple smoke test to make sure the index templates tab loads.

Fixes https://github.com/elastic/kibana/issues/27965
Fixes https://github.com/elastic/kibana/issues/19947

## Release notes
This feature adds the ability to create, edit and clone index templates within the Index Management app.

*Note*: This PR is continuation of https://github.com/elastic/kibana/pull/39922 and https://github.com/elastic/kibana/pull/41602.

## Screenshots

<details>
<summary>Template list and details with create, edit and clone actions added</summary>

<img width="1249" alt="Screen Shot 2019-08-16 at 9 10 13 PM" src="https://user-images.githubusercontent.com/5226211/63204909-fcb07580-c06a-11e9-920f-7732406fc494.png">

<img width="535" alt="Screen Shot 2019-08-16 at 9 10 27 PM" src="https://user-images.githubusercontent.com/5226211/63204911-01752980-c06b-11e9-81e6-dcd8d1e461d8.png">

</details>

<details>
<summary>Create wizard</summary>

Step 1:
<img width="995" alt="Screen Shot 2019-08-16 at 9 11 38 PM" src="https://user-images.githubusercontent.com/5226211/63204916-1782ea00-c06b-11e9-8be0-c8240059d561.png">

Step 1 with validation:
<img width="1247" alt="Screen Shot 2019-08-16 at 9 12 18 PM" src="https://user-images.githubusercontent.com/5226211/63204919-1ea9f800-c06b-11e9-9bd5-a7ae7a2c4bfb.png">

Step 2:
<img width="1243" alt="Screen Shot 2019-08-16 at 9 12 33 PM" src="https://user-images.githubusercontent.com/5226211/63204928-2f5a6e00-c06b-11e9-8223-c6e21858f619.png">

Step 2 with validation:
<img width="1245" alt="Screen Shot 2019-08-16 at 9 12 46 PM" src="https://user-images.githubusercontent.com/5226211/63204950-5fa20c80-c06b-11e9-8b17-b8f934e2f614.png">

Step 3:
<img width="1246" alt="Screen Shot 2019-08-16 at 9 12 57 PM" src="https://user-images.githubusercontent.com/5226211/63204932-3b463000-c06b-11e9-9feb-a982fa98f138.png">

Step 4:
<img width="1251" alt="Screen Shot 2019-08-16 at 9 13 04 PM" src="https://user-images.githubusercontent.com/5226211/63204934-40a37a80-c06b-11e9-932e-217f8ea426bb.png">

Step 5:
Summary tab
<img width="1236" alt="Screen Shot 2019-08-16 at 9 14 01 PM" src="https://user-images.githubusercontent.com/5226211/63204939-5153f080-c06b-11e9-92d3-d1a010c37668.png">

Request tab
<img width="1233" alt="Screen Shot 2019-08-16 at 9 14 16 PM" src="https://user-images.githubusercontent.com/5226211/63204941-55800e00-c06b-11e9-80db-abaeb2caf50c.png">

Wildcard warning
<img width="1230" alt="Screen Shot 2019-08-16 at 9 19 06 PM" src="https://user-images.githubusercontent.com/5226211/63204961-91b36e80-c06b-11e9-8d09-9a1cec8006aa.png">

</details>

<details>
<summary>Edit template</summary>

Name is read-only, but otherwise flow will be same as create.

<img width="1143" alt="Screen Shot 2019-08-16 at 9 20 30 PM" src="https://user-images.githubusercontent.com/5226211/63204992-dd661800-c06b-11e9-949b-6ff8b6805294.png">

</details>

<details>
<summary>Clone template</summary>

Clone will duplicate values from the selected template, with the exception of index patterns and aliases as per discussion with Yaron. A new name will also be created in the form of `<select_template_to_clone>-copy`. Otherwise, flow will be the same as create.

<img width="1113" alt="Screen Shot 2019-08-16 at 9 23 30 PM" src="https://user-images.githubusercontent.com/5226211/63205012-2027f000-c06c-11e9-91b4-d95016b912ab.png">


</details>

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10487/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10679/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10698/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11856/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11884/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11885/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12036/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12133/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12264/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12489/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12689/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12916/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12978/)

alisonelizabeth commented:
This is ready for an initial review!

FYI @silne30 if you'd like to review the tests.

Some things to note:
- I made the create/edit/clone flow a wizard rather than a one-page form, as in the original prototype, since the defined mappings could potentially be pretty long. I’m open to feedback if there are any concerns with this. I also have a design review scheduled later this week.
- When you clone a template, the index patterns and aliases are not copied over. This was suggested by Yaron, as it could be dangerous to have them copied over. 
- I originally had this integrated with the mappings editor implemented via https://github.com/elastic/kibana/pull/40415, but decided to remove it since it has not been fully tested and the latest work will not be released in `7.4`. 
- Working with Gail on copy changes; will address in a separate PR.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13057/)

alisonelizabeth commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13121/)

sebelga commented on x-pack/legacy/plugins/index_management/public/services/api.ts:
```diff
@@ -46,6 +49,16 @@ export const getHttpClient = () => {
   return httpClient;
 };
 
+let savedObjectsClient: any;
```
I would prefer the savedObjectClient to be in a separate file as it does not interact with the API layer. Also, see my comments about Kibana saved object "index patterns" to see if it is actually needed.

alisonelizabeth replied:
this has been removed

sebelga commented on x-pack/legacy/plugins/index_management/public/sections/home/template_list/template_details/template_details.tsx:
```diff
@@ -236,16 +281,58 @@ export const TemplateDetails: React.FunctionComponent<Props> = ({
+                      data-test-subj="manageTemplateButton"
+                      iconType="arrowDown"
+                      iconSide="right"
+                      onClick={() => setIsPopOverOpen(!isPopoverOpen)}
```
Info: it's a recommended practice to use the callBack when updating the state based on its previous value:

```js
onClick={() => setIsPopOverOpen(prev => !prev)}
```

alisonelizabeth replied:
good catch, fixed!

sebelga commented on x-pack/legacy/plugins/index_management/public/sections/template_create/template_create.tsx:
```diff
@@ -0,0 +1,166 @@
+/*
```
For the clone functionality, instead of having all the if / else statement I would have created a separate component. Don't you think it would be cleaner?

alisonelizabeth replied:
That's a valid point. I had considered that originally, but now that you bring it up again, I think that is the better approach. I went ahead and broke this up into a separate route and component.

sebelga replied:
Great, thanks!

sebelga commented on x-pack/legacy/plugins/index_management/public/services/api.ts:
```diff
@@ -201,19 +214,66 @@ export function loadIndexTemplates() {
+  return result;
+}
+
+export async function loadTemplateToClone(name: Template['name']) {
```
Nit: Why not call it simply `loadTemplate()`. As it could be used in a different context (not only cloning).

sebelga commented on x-pack/legacy/plugins/index_management/server/routes/api/templates/register_update_route.ts:
```diff
@@ -0,0 +1,37 @@
+  // Verify the template exists (ES will throw 404 if not)
+  await callWithRequest('indices.existsTemplate', { name });
+
+  // Otherwise update index template
```
Otherwise? 😊 Shouldn't it be "And then" ...

sebelga commented:
I also wanted to say: awesome test coverage! Great job 😊 ! @alisonelizabeth

alisonelizabeth commented on x-pack/legacy/plugins/index_management/common/lib/index.ts:
```diff
@@ -0,0 +1,10 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+export {
```
On the review step, there is the `JSON Request` tab, which shows JSON in the format as it would be sent to ES.

sebelga replied:
Got it 👍

cjcenizal replied:
@sebelga In general we will need serialization either client side or accessible to the client for the reason Alison mentioned, because our apps will expose the underlying Elasticsearch requests to the user wherever possible/useful.

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/components/template_form/steps/step_logistics.tsx:
```diff
@@ -0,0 +1,292 @@
+  const [allIndexPatterns, setAllIndexPatterns] = useState<Template['indexPatterns']>([]);
+
+  const getIndexPatterns = async () => {
+    const indexPatternObjects = await loadIndexPatterns();
```
@sebelga can you point in the right direction in the CCR code to take a look at this? The current implementation is the approach watcher uses to load index patterns when creating a threshold alert.

sebelga replied:
So the question is, you seem to be loading the Kibana index patterns. And if I am not wrong the patterns ES is expecting are different.

In CCR they are called "LeaderIndexPatterns"

https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/cross_cluster_replication/public/app/components/auto_follow_pattern_form.js#L121

alisonelizabeth replied:
@sebelga I understand, thanks.

It looks like CCR only allows custom values and then has an info callout when entered, showing potentially what kind of indices the pattern would match. What do you think if we follow the same pattern here?

<img width="623" alt="Screen Shot 2019-08-21 at 8 43 29 AM" src="https://user-images.githubusercontent.com/5226211/63432849-fa924200-c3ef-11e9-848f-2ee75ccade27.png">

sebelga replied:
Yes, it only allows custom values because patterns are custom by default 😊 (there is nowhere a user can save patterns in ES).

I am not sure that we need to add the preview, its purpose was mainly for the "prefix" and "suffix" that could be added.

alisonelizabeth replied:
Right, makes sense.

Fair enough :) Thanks for talking through this with me!

sebelga replied:
no prob!

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/components/template_form/steps/step_logistics.tsx:
```diff
@@ -0,0 +1,292 @@
+
+              const newIndexPatterns = [...indexPatterns, selectedPattern];
+
+              setAllIndexPatterns([...allIndexPatterns, selectedPattern]);
```
I was getting some weird behavior with the combo box when I didn't add it; it also seems to be how it is implemented in the EUI example (https://elastic.github.io/eui/#/forms/combo-box). But I might be missing something?

sebelga replied:
Let's see first if the index patterns suggestion is correct (below)

alisonelizabeth replied:
Yep, I think this is obsolete now.

Bamieh commented on x-pack/legacy/plugins/index_management/__jest__/client_integration/template_clone.test.tsx:
```diff
@@ -0,0 +1,159 @@
+      { attributes: { title: 'index2' } },
+      { attributes: { title: 'index3' } },
+    ];
+    return await INDEX_PATTERNS;
```
@sebelga 

using `return async` does have some newly introduced benefits in v8 engine used by Node.js, Chrome and a few other browsers:

v8 introduced a `--async-stack-traces` flag which as of V8 v7.3 is enabled by default (Node.js v12.0.0).

This flags provides an improved developer experience by enriching the Error stack property with async function calls stack trace.

    async function foo() {
      return bar();
    }
    
    async function bar() {
      await Promise.resolve();
      throw new Error('BEEP BEEP');
    }
    
    foo().catch(error => console.log(error.stack));


    Error: BEEP BEEP
        at bar (<anonymous>:7:9)

Note that by calling `return bar();` `foo()` function call does not appear at all in the error stack. Changing it to `return await bar();` gives a much better error stack output:

    async function foo() {
      return await bar();
    }

    Error: BEEP BEEP
        at bar (<anonymous>:7:9)
        at async foo (<anonymous>:2:10)

This indeed does provide much better error stack tracing, hence it is **HIGHLY encouraged to always await your promises.**

sebelga replied:
Thanks for the tip @Bamieh .

Although in this case, she is not returning a promise but a value. By creating an async function, the value (INDEX_PATTERNS) will be returned as a Promise.  Do you still think the `await` is needed?

alisonelizabeth replied:
Thanks for the explanation @Bamieh! Due to other changes in the code, this mock is no longer needed so it's been removed.

Bamieh commented on x-pack/legacy/plugins/index_management/public/components/template_form/template_steps.tsx:
```diff
@@ -0,0 +1,51 @@
+  maxCompletedStep,
+  updateCurrentStep,
+}) => {
+  const steps = [1, 2, 3, 4, 5].map(step => {
```
Maybe using `Object.entries(stepNamesMap).map(([step, stepName]) => ..` is better here?

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/components/template_form/template_form.tsx:
```diff
@@ -0,0 +1,202 @@
+    setTemplate(newTemplate);
+  };
+
+  const updateCurrentStep = (step: number) => {
```
I tried to simplify this a bit, and also made some changes to the functionality based on our other discussions as far as being able to freely proceed to different steps.

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/components/template_form/template_steps.tsx:
```diff
@@ -0,0 +1,51 @@
+      title: stepNamesMap[step],
+      isComplete: maxCompletedStep >= step,
+      isSelected: currentStep === step,
+      disabled: step === 1 ? false : maxCompletedStep < step - 1,
```
I changed the logic for this as described in my other comment.

alisonelizabeth commented:
@sebelga thanks for the review! It was very helpful!

I believe I addressed all of your concerns **except** relating to the tests. Since I made a fair number of changes, I wanted to get this out there as soon as possible to give you enough time to review again. I will address the tests in a separate commit.

In response to your questions:

> I find it strange that there is no way to return to the template list when we enter the "Edit" of a template. (apart from using the browser back button). Would it be complex to add another level in the breadcrumb (Management > Index Management > Index templates > Edit template) ?

I discussed this with design. Since index templates is a tab, not a page, they felt it didn't make sense to add another level to the breadcrumb. However, I did add a `Cancel` button throughout the wizard that will bring you back to the index templates tab.

> I found a bug, if you fill the step one, go to step 2. There you insert an **invalid** JSON data, click "next" (that will validate and display the JSON parse error). Then you click "back". The error is still there but you can't click "next" to correct it (and don't have an indication where to fix it).

Great catch! This should be fixed now, if you don't mind testing again behind me.

> I got the following error when trying to add aliases

Thanks for pointing this out! That is probably my mistake, the docs link originally pointed to `index aliases` but the format to define aliases is different within the template definition. I updated the docs link to point to the `Index templates` docs instead, and also added some help text with example JSON.

> A nice addon would be to validate the name when creating a new template (make sure it does not exist yet by making a debounced Http request). Also, it seems that `*` as a template name is invalid, trying to save a template with that name returns the following "There is already a template with name '*'." message. Even though there are no template (or system template) with that name.

There is already validation to make sure the template name doesn't already exist. I was reading through the API docs again, and the templates API accepts a wildcard for the GET request. Therefore, on the server, when we’re checking if the template already exists or not, the name `*` is returning true since it is matching all templates. I added validation on the client to not allow `*` at the end of the name. I also added validation for `,`, empty space, and `_`.

> Another validation that we could add is on **index patterns**. Spaces are not allowed for example (and other special characters, see auto_follow patterns). When I tried to save with a space in one of them, I got the following error which was not useful to understand it was related to index patterns.

There was already validation for invalid characters, but not spaces. I updated the logic for this to check for both. Thanks for pointing me to the auto_follow patterns for reference!

> In **edit** mode, I think all steps **should be enabled** by default as they all have already been filled. And we should be able to save the template on any step instead of having to click 4 times "next" to have the `save` button appear. This behavior could actually be present in **create** mode once the first step is filled (as the 3 others "settings", "mappings" and "aliases") are all optional. So after clicking "next" on step 1, all 5 steps are active + the save button is present.
> And for the **clone** functionality the same as for **edit**.
> What do you think?

I agree. Here's what I came up with: When you click `Next` or a different step via the steps component, validation will occur. If the step is invalid, you will not be taken to the next/future step, but you will be able to navigate back. Does that make sense? WDYT? If you don't mind testing behind me and making sure that works as expected that would be appreciated!

> I am not sure I see the value of the flyout when there is only the Summary tab as all the information and actions are available in the table. Shouldn't we keep all the tabs and display underneath something like: "This template does not have any mappings defined.".?

Good point! I ran this by design and they agreed. Updated the code.

> When I clone a template, the index patterns don't get copied over. Is this a bug?

This was intentional, but after receiving feedback, I made changes to the code so that everything is copied over on `Clone`.

alisonelizabeth commented:
FYI: In addition to updating the tests, I have a couple small items that I still need to address:
- [ ] Provide empty json object by default for mappings, settings and aliases editors
- [ ] Rename tabs in details panel when there is no info, e.g., `Mappings (none)`
- [ ] Add tooltip to `ILM policy` table header

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13657/)

sebelga commented on x-pack/legacy/plugins/index_management/public/components/template_form/template_form.tsx:
```diff
@@ -0,0 +1,273 @@
+    setTemplate(newTemplate);
+  };
+
+  const handleStepChange = (newCurrentStep: number, newMaxCompletedStep: number) => {
```
I still think the logic could still be simplified. Aside, there is a small bug. If we go to step 2, insert an invalid JSON then click "back".  And then go the review step, the invalid JSON is sent over.

<img width="250" alt="Screen Shot 2019-08-22 at 09 06 50" src="https://user-images.githubusercontent.com/2854616/63497398-f172a500-c4c3-11e9-997e-ca58765aee11.png">

This simplified logic fixes this bug

```js
const updateCurrentStep = (nextStep: number) => {
  // All the steps needs validation, except for the fifth
  const shouldValidate = currentStep !== 5;

  if (shouldValidate) {
    const stepValidation = validateStep(template);

    const { isValid: isCurrentStepValid } = stepValidation;

    const newValidation = {
      ...validation,
      ...{
        [currentStep]: stepValidation,
      },
    };

    setValidation(newValidation);

    // If step is invalid do not let user to proceed
    if (!isCurrentStepValid) {
      return;
    }
  }

  setCurrentStep(nextStep);
  setMaxCompletedStep(prev => Math.max(nextStep - 1, prev));
  clearSaveError();
};

const onBack = () => {
  updateCurrentStep(currentStep - 1);
};

const onNext = () => {
  updateCurrentStep(currentStep + 1);
};
```

alisonelizabeth replied:
good catch! the calculation for the max completed step differs for `onBack` vs `onNext`, so i tweaked this a little bit, but should be much cleaner now. thanks!

sebelga replied:
Mmmm... why does it defer?

alisonelizabeth replied:
Sorry, misspoke. The bug with the code above^ exists if you try to go **back > 1** step via the steps component. For example, if I'm on the `Aliases` step and click back to `Logistics`, this is what I see:

<img width="1299" alt="Screen Shot 2019-08-22 at 1 40 58 PM" src="https://user-images.githubusercontent.com/5226211/63537130-09a5ec80-c4e3-11e9-9a89-aceb61b20c65.png">

sebelga replied:
Ah I see. Then we don't need this `maxCompletedStep` state. If we completely remove it, then the <TemplateSteps /> component would have this JSX

```js
const steps = [1, 2, 3, 4, 5].map(step => {
    return {
      title: stepNamesMap[step],
      isComplete: currentStep > step,
      isSelected: currentStep === step,
      onClick: () => updateCurrentStep(step, step - 1),
    };
  });
```

sebelga commented on x-pack/legacy/plugins/index_management/public/components/template_form/template_steps.tsx:
```diff
@@ -0,0 +1,53 @@
+      title: stepNamesMap[step],
+      isComplete: maxCompletedStep >= step,
+      isSelected: currentStep === step,
+      disabled: isStepValid ? false : step > currentStep + 1,
```
I think this shoul be

```js
disabled: isStepValid ? false : step !== currentStep, // disable all step but the current one
onClick: () => updateCurrentStep(step), // no more "shouldValidate" needed from my previous comment
```

alisonelizabeth replied:
i was thinking about this some more - not sure that we really need to disable any of the steps, since we're always validating. WDYT? I removed it for now and also updated the `onClick`.

sebelga replied:
Yeah, both can work. The fact that they are disabled brings the focus to the form. We can ask another pair of eyes on it cc @cjcenizal

cjcenizal replied:
I'm not sure this works. If the first step is in an invalid state then I feel like all of the subsequent steps should be disabled, right?

![image](https://user-images.githubusercontent.com/1238659/63555436-bf7a3680-c4f5-11e9-88a5-dd91f441939d.png)

sebelga replied:
@cjcenizal  Yes, this is what

```js
disabled: isStepValid ? false : step !== currentStep, // disable all step but the current one
```

does. It disables all the steps once 1 step is invalid.

sebelga commented:
> I discussed this with design. Since index templates is a tab, not a page, they felt it didn't make sense to add another level to the breadcrumb. However, I did add a Cancel button throughout the wizard that will bring you back to the index templates tab.

The cancel button is great. I am not convinced by the design team decision. The user does not think in those terms (He does not ask: "_does the breadcrumb link point to a page or a tab in a page?_"). Aside, the cancel button might not be visible on all screen size.  

So as a user, this navigation is better (IMO):

<img width="1154" alt="Screen Shot 2019-08-22 at 10 23 44" src="https://user-images.githubusercontent.com/2854616/63498918-1d435a00-c4c7-11e9-8552-b2764c84ff14.png">

than this one (cancel button is not visible)

<img width="1157" alt="Screen Shot 2019-08-22 at 10 22 26" src="https://user-images.githubusercontent.com/2854616/63498948-29c7b280-c4c7-11e9-83d7-364e28c03ef3.png">

This is not a blocker, but I think it would improve the UX.

> There is already validation to make sure the template name doesn't already exist. 

I was referring to a client validation while giving a name to the template. Like we do in follower indices (https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/cross_cluster_replication/public/app/components/follower_index_form/follower_index_form.js#L168). This can be done in a second phase, especially with the Form lib that can do this much more easily.

> I added validation on the client to not allow * at the end of the name. I also added validation for ,, empty space, and _.

I played around with weird chars and got a few bugs. For example when using a `/` in the name (name: `"test/test")`. The redirect does not work (missing the encoding of the name). Then I gave name starting with a dot (name: `".\*t"`) and it appeared under the system indices (even though it was a user-generated one). Should we also prevent a name starting with a dot client-side and asking Elasticsearch to not allow it through the API? Using the escape char also has a strange behavior (name: `"\*\"`), but it seems related to the same non encoded redirect as with `/`.

> There was already validation for invalid characters, but not spaces. I updated the logic for this to check for both. 

I can still add values to the combo box with those invalid chars.

<img width="436" alt="Screen Shot 2019-08-22 at 10 52 41" src="https://user-images.githubusercontent.com/2854616/63500991-27ffee00-c4cb-11e9-9c11-4a8254b60f24.png">

Again, this will be very easily done with the Form hook goodies `<Field />` component (I will convince @cjcenizal of its benefit, I am sure of it! 😊 ). We can decide not to align behavior across our forms for now and wait for the form lib to be merged to ease those tasks as a second phase. But this proves the need for the lib to be merged asap.

alisonelizabeth commented on x-pack/legacy/plugins/index_management/public/sections/template_clone/template_clone.tsx:
```diff
@@ -0,0 +1,132 @@
+      ...{ name: `${name}-copy` },
+    } as Template;
+
+    content = renderTemplateForm(templateData);
```
If there is an error fetching the template to clone, a callout will display, but the create flow will not be blocked. In that scenario, the template form would take the default values (handled on line 103).

sebelga replied:
Right! I missed it 😊

alisonelizabeth commented:
@sebelga thanks for the second round of testing!

I think the steps behavior is in good shape now. Mind doing one last check?

I also added additional validation for the template name. The combo box/index patterns validation is done on submit. As a second phase, I think this can be improved as you mentioned above and we can incorporate the form lib.

Great point about the breadcrumb vs. cancel button. I chatted with CJ about this too and went ahead and added another layer as you suggested and removed the Cancel button. 

Still working on updating the tests...

sebelga commented on x-pack/legacy/plugins/index_management/public/components/template_form/template_form.tsx:
```diff
@@ -0,0 +1,245 @@
+
+  const onBack = () => {
+    const prevStep = currentStep - 1;
+    updateCurrentStep(prevStep, prevStep - 1);
```
The `maxCompletedStep` should not go down (always up). What was not working about my previous example?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13813/)

alisonelizabeth commented on x-pack/legacy/plugins/index_management/__jest__/client_integration/home.test.ts:
```diff
@@ -216,33 +223,59 @@ describe.skip('<IndexManagementHome />', () => {
+          expect(deleteAction.text()).toEqual('Delete');
+        });
+
+        test('template actions column should have an option to clone', async () => {
```
good catch! copy/paste mistake.

alisonelizabeth commented on x-pack/legacy/plugins/index_management/__jest__/client_integration/template_create.test.tsx:
```diff
@@ -0,0 +1,372 @@
+      { attributes: { title: 'index2' } },
+      { attributes: { title: 'index3' } },
+    ];
+    return await INDEX_PATTERNS;
```
no longer applicable

alisonelizabeth commented on x-pack/legacy/plugins/index_management/__jest__/client_integration/template_edit.test.tsx:
```diff
@@ -0,0 +1,170 @@
+    expect(find('pageTitle').text()).toEqual(`Edit template '${name}'`);
+  });
+
+  test('should have the correct props', async () => {
```
fair enough - removed!

alisonelizabeth commented on x-pack/legacy/plugins/index_management/__jest__/client_integration/template_clone.test.tsx:
```diff
@@ -0,0 +1,159 @@
+  // which does not produce a valid component wrapper
+  EuiComboBox: (props: any) => (
+    <input
+      data-test-subj="mockComboBox"
```
Good point. I'm using this in a couple different places, so I'm going to leave it as is for now, but may revisit later.

alisonelizabeth commented:
@sebelga tests has been updated!
<img width="740" alt="Screen Shot 2019-08-22 at 1 03 15 PM" src="https://user-images.githubusercontent.com/5226211/63536003-b7fc6280-c4e0-11e9-90a3-7a8df24e8d32.png">
<img width="766" alt="Screen Shot 2019-08-22 at 1 01 27 PM" src="https://user-images.githubusercontent.com/5226211/63536001-b7fc6280-c4e0-11e9-8537-13bafe7fca6b.png">
<img width="754" alt="Screen Shot 2019-08-22 at 1 00 51 PM" src="https://user-images.githubusercontent.com/5226211/63536000-b7fc6280-c4e0-11e9-8b2b-6749616b776f.png">
<img width="745" alt="Screen Shot 2019-08-22 at 1 02 07 PM" src="https://user-images.githubusercontent.com/5226211/63536002-b7fc6280-c4e0-11e9-98a4-bb224a92301e.png">

alisonelizabeth commented:
Thanks for the review @cjcenizal! I'll work on refactoring the validation/steps logic on Friday.

I believe I addressed all of your other comments though, with the exception of:

> A two-column layout seems like it would use this space a little more efficiently.

I originally had it this way, but design suggested I change it to one column - Caroline felt that it would make more use of the vertical space.

> I noticed that some of the columns could probably be made narrower to give more room to the name and index patterns columns.

I did have widths defined for the columns at one point, but noticed some alignment issues during IE testing. I'll have to go back and investigate some more, but likely will not be addressed in this PR.

sebelga commented:
@cjcenizal 
> If I enter invalid JSON on a later step and then try to go back to a previous step, I'm blocked.

I like this UX. Why let the user navigate away from an invalid form? Each step should be valid before going back/next or jump to another.

I think the rollup steps should not constrain other multi-step forms. In this concrete case, the only step that is required is the first one. A user needs to be able to jump from 2 to 4 or 5 if he wants IMO.

When Alison will make the change to update the validation `onChange`, this UX will be very the most flexible IMO

![demo_steps4](https://user-images.githubusercontent.com/2854616/63585440-6830a080-c59f-11e9-952a-e4ecbe341da9.gif)

alisonelizabeth commented:
I agree with @sebelga, I think previous steps should be disabled as well if the current step is invalid. 
Here's what it looks like now with my changes:

![templates_ui](https://user-images.githubusercontent.com/5226211/63595195-52fb4800-c586-11e9-9050-aac5449226db.gif)

alisonelizabeth commented:
Added logic to handle encoding/decoding special characters. It feels a little hacky to me; definitely would be interested in looking into a better long-term solution since this seems to be an issue across our apps.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14083/)

LeeDr commented:
Did anyone test this in Internet Explorer?  It seems completely broken in 7.4.0.

alisonelizabeth commented:
@LeeDr No worries. I did test this against IE11, but please let me know if you see any issues.

The table behavior is due to the fact that each row has >2 action types (edit, delete and clone). The EUI behavior is to show the `...` icon and only show 2 actions on hover. You can see a demo here: https://elastic.github.io/eui/#/display/tables. (You'll need to toggle the `Multiple actions` switch.)

Screenshot from EUI:
<img width="1001" alt="Screen Shot 2019-09-09 at 8 01 11 AM" src="https://user-images.githubusercontent.com/5226211/64529067-0a5ac300-d2d8-11e9-948d-a49fac55e998.png">

As far as the checkboxes, this is a pattern we use in many of the management apps. Typically, we've only allowed actions that can be performed in bulk. Let me know if you have any other questions!
